### PR TITLE
Feature/use implementation of iserializer

### PR DESCRIPTION
**Pull request by patheems**
Internal PR before requesting this on the orginal package.

This PR should solve the issue of being restricted to using Newtonsoft for serialization/deserialization.

### CODE REVIEW

File: src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs [hunk 1]
```diff
@@ -26,7 +27,7 @@ public void AddServices(IServiceCollection services)
             _options?.Invoke(dashboardOptions);
             services.AddTransient<IStartupFilter, CapStartupFilter>();
             services.AddSingleton(dashboardOptions);
-            services.AddSingleton(DashboardRoutes.Routes);
+            services.AddSingleton(x => new DashboardRoutes(x.GetRequiredService<ISerializer>()).Routes);
```
Comment (HBTeun) [inline src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs]: Maybe create a static method for this, so you don't have to create a new class instance

File: src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs [hunk 2]
```diff
@@ -56,10 +62,10 @@ public async Task<ConsumerExecutedResult> InvokeAsync(ConsumerContext context, C
                 else
                 {
                     if (message.Value != null)
-                    {
-                        if (message.Value is JToken jToken)  //reading from storage
+                    {
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs]: Remove the whitespace :)

File: src/DotNetCore.CAP/Messages/Message.cs [hunk 3]
```diff
@@ -9,16 +9,21 @@ namespace DotNetCore.CAP.Messages
...
+        public Message() {}
+
         public Message(IDictionary<string, string> headers, [CanBeNull] object value)
         {
             Headers = headers ?? throw new ArgumentNullException(nameof(headers));
             Value = value;
         }
 
-        public IDictionary<string, string> Headers { get; }
+        public IDictionary<string, string> Headers { get; set; }
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Messages/Message.cs:23]: I'm not sure a public set is wanted. The same for Value

File: src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs [hunk 4]
```diff
@@ -37,5 +38,25 @@ public Task<Message> DeserializeAsync(TransportMessage transportMessage, Type va
...
+        public object Deserialize(object value, Type valueType)
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool IsJsonType(object jsonObject)
+        {
+            return jsonObject is JToken;
+        }
+  }
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs:65]: Formatting

File: src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs [hunk 5]
```diff
@@ -25,7 +26,7 @@ public Task<TransportMessage> SerializeAsync(Message message)
 
             var json = JsonConvert.SerializeObject(message.Value);
             return Task.FromResult(new TransportMessage(message.Headers, Encoding.UTF8.GetBytes(json)));
-        }
+        }
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs]: Revert whitespace

File: src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs [hunk 6]
```diff
@@ -37,5 +38,25 @@ public Task<Message> DeserializeAsync(TransportMessage transportMessage, Type va
             var json = Encoding.UTF8.GetString(transportMessage.Body);
             return Task.FromResult(new Message(transportMessage.Headers, JsonConvert.DeserializeObject(json, valueType)));
         }
-    }
+
+        public string Serialize(Message message)
+        {
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs]: You should create a working version. So no NotImplementedExceptions

File: src/DotNetCore.CAP/Serialization/ISerializer.cs [hunk 7]
```diff
@@ -16,8 +16,38 @@ public interface ISerializer
...
+        object Deserialize(object value, Type valueType);
+
+        /// <summary>
+        /// Check if the given object is of Json type, e.g. JToken or JsonElement
+        /// depending on the type of serializer implemented
+        /// </summary>
+        /// <example>
+        /// <code>
+        /// // Example implementation for System.Text.Json
+        ///  public bool IsJsonType(object jsonObject)
```
Comment (HBTeun) [inline src/DotNetCore.CAP/Serialization/ISerializer.cs:45]: Remove the added space

File: src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs (hunk 6 above)
Comment (HBTeun) [reply src/DotNetCore.CAP/Serialization/ISerializer.JsonUtf8.cs]: Run the tests. This should be shown in the tests

**Review by HBTeun — CHANGES_REQUESTED** [review]

File: src/DotNetCore.CAP.InMemoryStorage/IDataStorage.InMemory.cs [hunk 8]
```diff
@@ -12,16 +12,23 @@
...
 {
     internal class InMemoryStorage : IDataStorage
     {
+        private readonly IServiceProvider _serviceProvider;
+
         private readonly IOptions<CapOptions> _capOptions;
+        private readonly ISerializer _serializer;
 
-        public InMemoryStorage(IOptions<CapOptions> capOptions)
+        public InMemoryStorage(IOptions<CapOptions> capOptions, IServiceProvider serviceProvider)
```
Comment (Mastermindzh) [inline src/DotNetCore.CAP.InMemoryStorage/IDataStorage.InMemory.cs]: Why not just inject the ISerializer?

File: src/DotNetCore.CAP.MongoDB/IDataStorage.MongoDB.cs [hunk 9]
```diff
@@ -27,8 +31,12 @@ public class MongoDBDataStorage : IDataStorage
         public MongoDBDataStorage(
             IOptions<CapOptions> capOptions,
             IOptions<MongoDBOptions> options,
-            IMongoClient client)
+            IMongoClient client,
+            IServiceProvider serviceProvider)
```
Comment (Mastermindzh) [inline src/DotNetCore.CAP.MongoDB/IDataStorage.MongoDB.cs]: Why not just inject the ISerializer here as well?

File: src/DotNetCore.CAP.MySql/IDataStorage.MySql.cs [hunk 10]
```diff
@@ -12,24 +12,33 @@
...
+
         private readonly string _pubName;
         private readonly string _recName;
 
         public MySqlDataStorage(
             IOptions<MySqlOptions> options,
             IOptions<CapOptions> capOptions,
-            IStorageInitializer initializer)
+            IStorageInitializer initializer,
+            IServiceProvider serviceProvider)
```
Comment (Mastermindzh) [inline src/DotNetCore.CAP.MySql/IDataStorage.MySql.cs]: Same!

File: src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs (hunk 2 above)
Comment (Mastermindzh) [reply src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs]: Even on github! I commend your steadfastness!

File: src/DotNetCore.CAP/Messages/Message.cs (hunk 3 above)
Comment (Mastermindzh) [reply src/DotNetCore.CAP/Messages/Message.cs:23]: The system.text.json serializer requires a parameterless constructor and seems to skip properties without a set;

That was the problem we faced when we changed this to set;

Any better ideas?

**Review by Mastermindzh — APPROVED** [review]
Great work @patheems !

Really glad we could solve this and integrate the ISerializer in the entire product.

File: src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs (hunk 1 above)
Comment (patheems) [reply src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs]: Any helpful ideas would be nice. The original code is using static but due to the need of the ISerializer this was changed.

File: src/DotNetCore.CAP.InMemoryStorage/IDataStorage.InMemory.cs (hunk 8 above)
Comment (patheems) [reply src/DotNetCore.CAP.InMemoryStorage/IDataStorage.InMemory.cs]: Changed to suggestion.

File: src/DotNetCore.CAP.MongoDB/IDataStorage.MongoDB.cs (hunk 9 above)
Comment (patheems) [reply src/DotNetCore.CAP.MongoDB/IDataStorage.MongoDB.cs]: Done

File: src/DotNetCore.CAP.MySql/IDataStorage.MySql.cs (hunk 10 above)
Comment (patheems) [reply src/DotNetCore.CAP.MySql/IDataStorage.MySql.cs]: Done.

File: src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs (hunk 2 above)
Comment (patheems) [reply src/DotNetCore.CAP/Internal/ISubscribeInvoker.Default.cs]: Should be done...

**Review by Sharonfranke — APPROVED** [review]

File: src/DotNetCore.CAP/Messages/Message.cs (hunk 3 above)
Comment (HBTeun) [reply src/DotNetCore.CAP/Messages/Message.cs:23]: I'd forgotten that. 
But you could create a[ custom serializer](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to) for the class.
(or switch to .net 5.0, it's possible there)

File: src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs (hunk 1 above)
Comment (HBTeun) [reply src/DotNetCore.CAP.Dashboard/CAP.DashboardOptionsExtensions.cs]: The constructor is already a method just rename and add the return value, f.e.:
public List<string> DashboardRoutes.GetRoutes(ISerializer serializer)

**Comment by patheems** [comment]
Already merged, obsolete

**Pull request closed without merging**